Microsoft.Toolkit.Uwp.UI.Media/Geometry/Utils.cs

```diff
@@ -0,0 +1,905 @@
+            // If they're both empty, don't bother with the double logic.
+            if (rect1.IsEmpty)
+            {
+                return rect2.IsEmpty;
+            }
+
+            // At this point, rect1 isn't empty, so the first thing we can test is rect2.IsEmpty, followed by property-wise compares.
+            return (!rect2.IsEmpty)
+                   && rect1.X.IsCloseTo(rect2.X) && rect1.Y.IsCloseTo(rect2.Y)
+                   && rect1.Height.IsCloseTo(rect2.Height) && rect1.Width.IsCloseTo(rect2.Width);
+        }
+
+        /// <summary>
+        /// Restricts the given single precision number to be within the
+        /// specified min and max values (inclusive).
+        /// </summary>
+        /// <param name="num">Given single precision number</param>
+        /// <param name="min">Minimum value</param>
+        /// <param name="max">Maximum value</param>
+        /// <returns>Clamped value</returns>
+        public static float Clamp(this float num, float min, float max)
+        {
+            if (min > max)
+            {
+                var temp = min;
+                min = max;
+                max = temp;
+            }
+
+            return Math.Max(min, Math.Min(max, num));
+        }
+
+        /// <summary>
+        /// Restricts the given double precision number to be within the
+        /// specified min and max values (inclusive).
+        /// </summary>
+        /// <param name="num">Given double precision number</param>
+        /// <param name="min">Minimum value</param>
+        /// <param name="max">Maximum value</param>
+        /// <returns>Clamped value</returns>
+        public static double Clamp(this double num, double min, double max)
+        {
+            if (min > max)
+            {
+                var temp = min;
+                min = max;
+                max = temp;
+            }
+
+            return Math.Max(min, Math.Min(max, num));
+        }
+
+        [StructLayout(LayoutKind.Explicit)]
+        private struct NanUnion
+        {
+            [FieldOffset(0)]
+            public double DoubleValue;
+            [FieldOffset(0)]
+            public ulong UintValue;
+        }
+
+        /// <summary>
+        /// Faster check for NaN ( faster than double.IsNaN() )
+        /// IEEE 754 : If the argument is any value in the range 0x7ff0000000000001L through 0x7fffffffffffffffL
+        /// or in the range 0xfff0000000000001L through 0xffffffffffffffffL, the result will be NaN.
+        /// </summary>
+        /// <param name="value">Value to check</param>
+        /// <returns>True/False</returns>
+        public static bool IsNaN(double value)
+        {
+            var t = new NanUnion
+            {
+                DoubleValue = value
+            };
+
+            var exp = t.UintValue & 0xfff0000000000000;
+            var man = t.UintValue & 0x000fffffffffffff;
+
+            return (exp == 0x7ff0000000000000 || exp == 0xfff0000000000000) && (man != 0);
+        }
```

Comment:
Is this actually needed, instead of just using the spec conformant API from the BCL?

There are all sorts of issues that can arise in this area with different parts of a framework using slightly different behaviors to handle NaN-s and whatnot for IEEE-compliant floating point values. Given the rest of the codebase I'd really be surprised if this made a noticeable difference. Also this actually seems much slower than the official `IsNaN` from the BCL, see [here](https://github.com/dotnet/runtime/blob/5078dd61a81c88dbdd131a77b8148b4dee3803fa/src/libraries/System.Private.CoreLib/src/System/Double.cs#L91). This looks like a workaround for a previous `IsNaN` implementation that doesn't seem to be there anymore, shouldn't we just remove this and use the one from the BCL now?

Reply:
Removed `IsNaN()` method from Utils. Using `double.IsNaN` instead.